agent/dockerclient/dockerapi/docker_client.go

```diff
@@ -837,36 +838,55 @@ func getMetadataHealthCheck(dockerContainer *types.ContainerJSON) apicontainer.H
 
 // Listen to the docker event stream for container changes and pass them up
 func (dg *dockerGoClient) ContainerEvents(ctx context.Context) (<-chan DockerContainerChangeEvent, error) {
-	client, err := dg.dockerClient()
+	client, err := dg.sdkDockerClient()
 	if err != nil {
 		return nil, err
 	}
-	dockerEvents := make(chan *docker.APIEvents, dockerEventBufferSize)
-	events := make(chan *docker.APIEvents)
+
+	events := make(chan *events.Message)
 	buffer := NewInfiniteBuffer()
 
-	err = client.AddEventListener(dockerEvents)
-	if err != nil {
-		seelog.Errorf("DockerGoClient: unable to add a docker event listener: %v", err)
-		return nil, err
-	}
+	derivedCtx, cancel := context.WithCancel(ctx)
+	dockerEvents, eventErr := client.Events(derivedCtx, types.EventsOptions{})
+
+	// Cache the event from docker client. Channel closes when an error is passed to eventErr.
+	go buffer.StartListening(derivedCtx, dockerEvents)
+	// Receive errors from channels. If error thrown is not EOF, log and reopen channel.
 	go func() {
-		<-ctx.Done()
-		client.RemoveEventListener(dockerEvents)
+		for {
+			select {
+			case err := <-eventErr:
+				if err == io.EOF {
+					// If error processed is EOF, ignore and continue listening
```

Comment:
not sure why we ended up like this in previous pr, but from the implementation of client.Events(https://github.com/docker/go-docker/blob/4daae26030ad00e348edddff9767924ae57a3b82/events.go#L20), it seems that after an error comes out from the error channel (not matter it's EOF or not), there will be no more events coming out from event channel, and we need to call client.Events again to continue receiving event